### PR TITLE
IC-1846 / IC-1847 store timestamp when updating draft referral risk information / update column names

### PR DESCRIPTION
**Pull request by tomsmyers**
## What does this pull request do?

IC-1846 store timestamp when updating draft referral risk information
IC-1847 update database column names with more descriptive field names (but don't change the API)

## What is the intent behind these changes?

for auditing purposes we know exactly when the PP entered the risk information - even if it's quite different from when they sent off the referral

make the data dictionary clear in terms of what the risk fields represent.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt [hunk 1]
```diff
@@ -161,7 +161,7 @@ class ReferralService(
       referral.id,
       referral.serviceUserCRN,
       user,
-      OffsetDateTime.now(), // fixme: this should be the timestamp which we store when additionalRiskInformation is set
+      referral.additionalRiskInformationUpdatedAt,
       riskInformation,
     )
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:167]: should we clear the timestamp below too??

Comment (sldblog) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:167]: would be good to retain it until we have a way to retrieve it from ARN

Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:167]: 👍🏼

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/RisksAndNeedsService.kt [hunk 2]
```diff
@@ -22,7 +22,7 @@ data class CreateSupplementaryRiskRequest(
   val crn: String,
   val createdByUser: String,
   val createdByUserType: String,
-  val createdDate: LocalDateTime,
+  val createdDate: LocalDateTime?,
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/RisksAndNeedsService.kt:25]: It sounds like this field shouldn't be optional -- intent-wise there must always be a time where the risk is 'signalled'

What do you think about setting `additionalRiskInformationUpdatedAt` to current time in `ReferralService.submitAdditionalRiskInformation` if it's missing? (Together with a warning?)

**Review by sldblog — COMMENTED** [review]
👍 nice -- just a question about `riskCreatedAt` field being optional

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/RisksAndNeedsService.kt (hunk 2 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/RisksAndNeedsService.kt:25]: yes happy with that idea.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Referral.kt [hunk 3]
```diff
@@ -54,7 +54,8 @@ class Referral(
 
   // draft referral fields
   @OneToOne(mappedBy = "referral", cascade = arrayOf(CascadeType.ALL)) @PrimaryKeyJoinColumn var serviceUserData: ServiceUserData? = null,
-  var additionalRiskInformation: String? = null,
+  @Column(name = "draft_supplementary_risk") var additionalRiskInformation: String? = null,
```
[the remark this answers is not in the document]
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/Referral.kt:57]: i tried doing this. and i found it confusing having the difference there between the entity field and the DTO field - especially in the test code.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt [hunk 4]
```diff
@@ -157,11 +157,17 @@ class ReferralService(
     val riskInformation = referral.additionalRiskInformation
       ?: throw ServerWebInputException("can't submit a referral without risk information")
 
+    val riskSubmittedAt = referral.additionalRiskInformationUpdatedAt
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt:160]: @sldblog changes as suggested here

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralServiceUnitTest.kt [hunk 5]
```diff
@@ -578,41 +589,69 @@ class ReferralServiceUnitTest {
...
 
-    @Test
-    fun `cant send draft referral without risk information`() {
-      val referral = referralFactory.createDraft()
-      val authUser = authUserFactory.create()
-
-      val e = assertThrows<ServerWebInputException> {
-        referralService.sendDraftReferral(referral, authUser)
-      }
+  @Test
```
Comment (sldblog) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralServiceUnitTest.kt:594]: _nit_ Wondering if we should have a ktlint job 🤔

**Review by sldblog — APPROVED** [review]
🎉

**Pull request merged**